server/src/test/java/org/elasticsearch/common/util/concurrent/AbstractThrottledTaskRunnerTests.java

```diff
@@ -143,6 +143,52 @@ public void onResponse(Releasable releasable) {
         assertNoRunningTasks(taskRunner);
     }
 
+    public void testRunSyncTasksEagerly() {
+        final int maxTasks = randomIntBetween(1, maxThreads);
+        final int taskCount = between(maxTasks, maxTasks * 2);
+        final var barrier = new CyclicBarrier(maxTasks + 1);
+        final var executedCountDown = new CountDownLatch(taskCount);
+        final var testThread = Thread.currentThread();
+
+        class TestTask implements ActionListener<Releasable> {
+
+            @Override
+            public void onFailure(Exception e) {
+                throw new AssertionError(e);
+            }
+
+            @Override
+            public void onResponse(Releasable releasable) {
+                try (releasable) {
+                    if (Thread.currentThread() != testThread) {
+                        safeAwait(barrier);
+                        safeAwait(barrier);
+                    }
+                } finally {
+                    executedCountDown.countDown();
+                }
+            }
+        }
+
+        final BlockingQueue<TestTask> queue = ConcurrentCollections.newBlockingQueue();
+        final AbstractThrottledTaskRunner<TestTask> taskRunner = new AbstractThrottledTaskRunner<>("test", maxTasks, executor, queue);
+        for (int i = 0; i < taskCount; i++) {
+            taskRunner.enqueueTask(new TestTask());
+        }
+
+        safeAwait(barrier);
+        assertEquals(taskCount - maxTasks, queue.size());
+        assertThat(taskRunner.runningTasks(), equalTo(maxTasks));
+
+        taskRunner.runSyncTasksEagerly();
+        assertEquals(0, queue.size());
+
+        safeAwait(barrier);
+        safeAwait(executedCountDown);
+        assertTrue(queue.isEmpty());
+        assertNoRunningTasks(taskRunner);
```

Comment:
For my learning:
1. Why do we need assert the queue is empty here? We already asserted it has 0 size four lines above. I don't see how the queue size can increase after all tasks are enqueued?
2. In the existing `assertNoRunningTasks`, why is it necessary to spawn a batch of `Runnable`s before verify `runningTasks` size is 0? We already verified that the `queue` is empty. It seems to me that just verify `runningTasks` is sufficient? Also running the extra `Runnable` does not guarantee that they actually excercise every thread in the pool. So not sure the purpose here.

Reply:
Sorry for dragging on this point. But I think this threads flushing behaviour is not necessary for this test because the `releasable` is used in a try-with-resource block and should be released before `executedCountDown` finishes. This behaviour is needed for `testEnqueueSpawnsNewTasksUpToMax` because it [explicitly closes the releasable after the countDown](https://github.com/elastic/elasticsearch/blob/055df9a104a646cbb12db0b4d11ccaf6ecba6ffa/server/src/test/java/org/elasticsearch/common/util/concurrent/AbstractThrottledTaskRunnerTests.java#L118-L119). If I flip the order, it also runs successfully 10K times without the thread flushing logic.

Reply:
Ah hm I see what you mean now. Yes I think you're right we could simplify this if we made sure to release all the task refs before allowing that test to proceed.